Running StreamConsumer with tokio's executor

Currently how StreamConsumer stores the `JoinHandle` in a `Cell` prevents us of using the consumer in tokio's [thread_pool](https://docs.rs/tokio/0.1.6/tokio/executor/thread_pool/index.html) executor, which expects the consumer to implement `Send`. In other hand using the [current_thread](https://docs.rs/tokio/0.1.6/tokio/executor/current_thread/index.html) executor doesn't comply with missing static lifetimes somewhere in the consumer code I haven't been able to locate yet.

Is there plans to support the new executors, especially when the development of `tokio-core` and `futures-cpupool` seems to be stalled in favor of `tokio`?

Tried to fix this and hit the wall, my try is here so you'll see the approach:

https://github.com/fede1024/rust-rdkafka/pull/87

The error I get is 
```
error[E0277]: the trait bound `*mut rdkafka_sys::bindings::linux_64::rd_kafka_message_s: std::marker::Send` is not satisfied in `futures::Then<futures::Lazy<[closure@examples/asynchronous_processing.rs:85:39: 107:14 borrowed_message:rdkafka::message::BorrowedMessage, producer:rdkafka::producer::FutureProducer, output_topic:std::string::String], futures::AndThen<futures::Lazy<[closure@examples/asynchronous_processing.rs:87:40: 90:20 borrowed_message:rdkafka::message::BorrowedMessage], futures::future::PollFn<[closure@examples/asynchronous_processing.rs:87:56: 90:19 borrowed_message:rdkafka::message::BorrowedMessage]>>, futures::Then<rdkafka::producer::DeliveryFuture, std::result::Result<(), tokio_threadpool::BlockingError>, [closure@examples/asynchronous_processing.rs:98:31: 105:26]>,
[closure@examples/asynchronous_processing.rs:92:38: 106:18 producer:rdkafka::producer::FutureProducer, output_topic:std::string::String]>>, futures::FutureResult<(), ()>, [closure@examples/asynchronous_processing.rs:109:46: 109:56]>`
   --> examples/asynchronous_processing.rs:109:13
    |
109 |             tokio::spawn(message_future.then(|_| ok(())));
    |             ^^^^^^^^^^^^ `*mut rdkafka_sys::bindings::linux_64::rd_kafka_message_s` cannot be sent between threads safely
    |
```

Also removing the lifetimes with the consumers while having unsafe and raw pointers around doesn't feel very safe. Could we actually have a safe wrappers here so doing this kind of refactorings would be easier?

How is this workaround supposed to work? I'm trying to use it on the `run_async_processor` example so:

```
let owned_stream: OwnedConsumerStream = OwnedConsumerStream {
        upstream: OwningHandle::new_with_fn(
            Box::new(consumer)
            ,
              |cons| {
                 cons.start()
             })
    };
```
but I get 

`error[E0599]: no method named `start` found for type `*const rdkafka::consumer::StreamConsumer` in the current scope`

if this issue is not fixed, this library becomes unusable. core.run(f) is deprecated and when i try doing
tokio::run(h), it gives me below error
error[E0277]: `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>` cannot be shared between threads safely
    --> src/main.rs:1836:5                                                                                                                                                
     |                                                                                                                                                                    
1836 |     tokio::run(h);                                                                                                                                                 
     |     ^^^^^^^^^^ `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>` cannot be shared between threads safely                                         
     |                                                                                                                                                                    
     = help: within `rdkafka::consumer::StreamConsumer`, the trait `std::marker::Sync` is not implemented for `std::cell::Cell<std::option::Option<std::thread::JoinHandle<()>>>`
     = note: required because it appears within the type `rdkafka::consumer::StreamConsumer`                                                                              
     = note: required because of the requirements on the impl of `std::marker::Send` for `&rdkafka::consumer::StreamConsumer`                                             
     = note: required because it appears within the type `rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>`                                
     = note: required because it appears within the type `futures::stream::FilterMap<rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>, [closure@src/main.rs:1755:42: 1763:10]>`
     = note: required because it appears within the type `futures::stream::ForEach<futures::stream::FilterMap<rdkafka::consumer::MessageStream<'_, rdkafka::consumer::DefaultConsumerContext>, [closure@src/main.rs:1755:42: 1763:10]>, [closure@src/main.rs:1763:21: 1829:10 consumer:_, tx:_, input_topic:_], futures::AndThen<futures::MapErr<futures::Receiver<std::string::String>, [closure@src/main.rs:1802:20: 1807:13]>, std::result::Result<(), ()>, [closure@src/main.rs:1813:19: 1826:10 k:_, t1:_]>>`
     = note: required by `tokio::run`

--- a/examples/asynchronous_processing.rs
+++ b/examples/asynchronous_processing.rs
@@ -1,20 +1,27 @@
 use std::thread;
 use std::time::Duration;
 
-use clap::{App, Arg};
-use futures::{future, TryStreamExt};
-use log::{info, warn};
+use clap::{App, Arg, value_t};
+use futures::{StreamExt, TryStreamExt};
+use futures::stream::FuturesUnordered;
+use log::info;
 
 use rdkafka::config::ClientConfig;
 use rdkafka::consumer::stream_consumer::StreamConsumer;
 use rdkafka::consumer::Consumer;
-use rdkafka::message::OwnedMessage;
+use rdkafka::message::{BorrowedMessage, OwnedMessage};
 use rdkafka::producer::{FutureProducer, FutureRecord};
 use rdkafka::Message;
 
 use crate::example_utils::setup_logger;
 
 mod example_utils;
+
+async fn record_message_receipt(msg: &BorrowedMessage<'_>) {
+    // Simulate some work that must be done in the same order as messages are
+    // received; i.e., before truly parallel processing can begin.
+    info!("Message received: {}", msg.offset());
+}
 
 // Emulates an expensive, synchronous computation.
 fn expensive_computation<'a>(msg: OwnedMessage) -> String {
@@ -39,11 +46,16 @@ fn expensive_computation<'a>(msg: OwnedMessage) -> String {
 // `tokio::spawn` is used to handle IO-bound tasks in parallel (e.g., producing
 // the messages), while `tokio::task::spawn_blocking` is used to handle the
 // simulated CPU-bound task.
-async fn run_async_processor(brokers: &str, group_id: &str, input_topic: &str, output_topic: &str) {
+async fn run_async_processor(
+    brokers: String,
+    group_id: String,
+    input_topic: String,
+    output_topic: String,
+) {
     // Create the `StreamConsumer`, to receive the messages from the topic in form of a `Stream`.
     let consumer: StreamConsumer = ClientConfig::new()
-        .set("group.id", group_id)
-        .set("bootstrap.servers", brokers)
+        .set("group.id", &group_id)
+        .set("bootstrap.servers", &brokers)
         .set("enable.partition.eof", "false")
         .set("session.timeout.ms", "6000")
         .set("enable.auto.commit", "false")
@@ -51,46 +63,48 @@ async fn run_async_processor(brokers: &str, group_id: &str, input_topic: &str, o
         .expect("Consumer creation failed");
 
     consumer
-        .subscribe(&[input_topic])
+        .subscribe(&[&input_topic])
         .expect("Can't subscribe to specified topic");
 
     // Create the `FutureProducer` to produce asynchronously.
     let producer: FutureProducer = ClientConfig::new()
-        .set("bootstrap.servers", brokers)
+        .set("bootstrap.servers", &brokers)
         .set("message.timeout.ms", "5000")
         .create()
         .expect("Producer creation error");
 
     // Create the outer pipeline on the message stream.
     let stream_processor = consumer.start().try_for_each(|borrowed_message| {
-        // Process each message
-        info!("Message received: {}", borrowed_message.offset());
-        // Borrowed messages can't outlive the consumer they are received from, so they need to
-        // be owned in order to be sent to a separate thread.
-        let owned_message = borrowed_message.detach();
-        let output_topic = output_topic.to_string();
         let producer = producer.clone();
-        tokio::spawn(async move {
-            // The body of this block will be executed on the main thread pool,
-            // but we perform `expensive_computation` on a separate thread pool
-            // for CPU-intensive tasks via `tokio::task::spawn_blocking`.
-            let computation_result =
-                tokio::task::spawn_blocking(|| expensive_computation(owned_message))
-                    .await
-                    .expect("failed to wait for expensive computation");
-            let produce_future = producer.send(
-                FutureRecord::to(&output_topic)
-                    .key("some key")
-                    .payload(&computation_result),
-                0,
-            );
-            match produce_future.await {
-                Ok(Ok(delivery)) => println!("Sent: {:?}", delivery),
-                Ok(Err((e, _))) => println!("Error: {:?}", e),
-                Err(_) => println!("Future cancelled"),
-            }
-        });
-        future::ready(Ok(()))
+        let output_topic = output_topic.to_string();
+        async move {
+            // Process each message
+            record_message_receipt(&borrowed_message).await;
+            // Borrowed messages can't outlive the consumer they are received from, so they need to
+            // be owned in order to be sent to a separate thread.
+            let owned_message = borrowed_message.detach();
+            tokio::spawn(async move {
+                // The body of this block will be executed on the main thread pool,
+                // but we perform `expensive_computation` on a separate thread pool
+                // for CPU-intensive tasks via `tokio::task::spawn_blocking`.
+                let computation_result =
+                    tokio::task::spawn_blocking(|| expensive_computation(owned_message))
+                        .await
+                        .expect("failed to wait for expensive computation");
+                let produce_future = producer.send(
+                    FutureRecord::to(&output_topic)
+                        .key("some key")
+                        .payload(&computation_result),
+                    0,
+                );
+                match produce_future.await {
+                    Ok(Ok(delivery)) => println!("Sent: {:?}", delivery),
+                    Ok(Err((e, _))) => println!("Error: {:?}", e),
+                    Err(_) => println!("Future cancelled"),
+                }
+            });
+            Ok(())
+        }
     });
 
     info!("Starting event loop");
@@ -139,6 +153,13 @@ async fn main() {
                 .takes_value(true)
                 .required(true),
         )
+        .arg(
+            Arg::with_name("num-workers")
+                .long("num-workers")
+                .help("Number of workers")
+                .takes_value(true)
+                .default_value("1"),
+        )
         .get_matches();
 
     setup_logger(true, matches.value_of("log-conf"));
@@ -147,6 +168,18 @@ async fn main() {
     let group_id = matches.value_of("group-id").unwrap();
     let input_topic = matches.value_of("input-topic").unwrap();
     let output_topic = matches.value_of("output-topic").unwrap();
+    let num_workers = value_t!(matches, "num-workers", usize).unwrap();
 
-    run_async_processor(brokers, group_id, input_topic, output_topic).await
+    (0..num_workers)
+        .map(|_| {
+            tokio::spawn(run_async_processor(
+                brokers.to_owned(),
+                group_id.to_owned(),
+                input_topic.to_owned(),
+                output_topic.to_owned(),
+            ))
+        })
+        .collect::<FuturesUnordered<_>>()
+        .for_each(|_| async { () })
+        .await
 }

--- a/src/message.rs
+++ b/src/message.rs
@@ -362,6 +362,9 @@ impl<'a> Drop for BorrowedMessage<'a> {
     }
 }
 
+unsafe impl<'a> Send for BorrowedMessage<'a> {}
+unsafe impl<'a> Sync for BorrowedMessage<'a> {}
+
 //
 // ********** OWNED MESSAGE **********
 //

